doc/compiler/AOTClassChains.md

```diff
@@ -0,0 +1,153 @@
+then `J9Class(C'') == J9Class(C')`. This means that objects of `C''` or `C'` 
+have the same shape, and any relationships known between `J9Class(C')` and
+`J9Class(B')` are also true between `J9Class(C'')` and `J9Class(B'')`.
+
+Class Chains are this set of romclass relationships, and they are stored 
+in the SCC. They are a list of romclasses; one first walks up the super 
+class chain up to `java/lang/Object`, and then walks the class' iTable 
+to get the full list of interface classes that are implemented. During 
+a compilation, the JIT creates the list in the order described above 
+and stores it into the SCC. Each romclass is encoded as an offset from 
+the beginning of the SCC. The offsets of various class chains in the SCC 
+are referenced in the various Relocation Records, which describe the list 
+of `J9Class`es an AOT compilation has asked questions about. 
+
+When AOT code, generated in one JVM instance, is loaded into a different 
+JVM instance, the AOT infrastructure processes the Relocation Records; 
+it finds a candidate `J9Class` pointer in the current JVM and walks its 
+superclasses and interfaces, checking each one's romclass against the one 
+in the Class Chain stored in the SCC. Any mismatch implies the classes 
+might be different, in which case the AOT load fails. In the case of 
+inlined methods however, the AOT load doesn't necessarily fail, rather, 
+the AOT infrastructure disables a particular inlined call site.
+
+## Finding a candidate J9Class
+Finding a candidate J9Class with which to perform the validation is 
+non-trivial. Java uses Class Loaders to load classes. Two class loaders 
+can load the same "C extends B extends A" class twice, and this will 
+result in two different `J9Class` pointers; the difference comes from 
+the fact the two class loaders are different. Frameworks such as OSGI 
+uses different class loaders for different modules, and Java 
+applications can use several OSGI modules. Even if there weren't 
+multiple instances of a particular class in the current JVM instance, 
+having to go through every possible class loader to check if it already 
+has that particular class loaded would be extremely slow. If there were 
+multiple instances, then the AOT infrastructure would additionally need 
+to validate every single one. All this would effectively defeat the 
+purpose of AOT code, which is to facilitate fast application startup.
+
+Finding a particular `J9Class` in the current JVM instance requires 
+a Class Loader and a Class Name. The Class Name can be obtained from 
+the top romclass in the class chain. However, finding the Class Loader 
+is not as simple since they are just Java Objects created by a Java 
+application. OpenJ9 solves this problem by building has tables that 
+map a) Class Loaders to the first Class that loader loaded, and b) the 
+first class that loader loaded to the Class Loader Object. These tables
```

Comment:
@mstoodle To clarify, regarding the "first class that the loader loaded", what gets stored is the romclass (or its offset) right?

Reply:
what gets stored is an offset to the class chain for that class...